### PR TITLE
configstate,hookstate: timeout the configure hook after 5 mins, report failures to the errtracker

### PR DESCRIPTION
**Pull request by mvo5**
Build on top of https://github.com/snapcore/snapd/pull/3074

Easy review via https://github.com/mvo5/snappy/compare/hook-time-limit...mvo5:hook-time-limit-and-errtracker

### CODE REVIEW

File: overlord/hookstate/hookmgr.go [hunk 1]
```diff
@@ -296,16 +332,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+		return buffer.Bytes(), abortOrTimeoutError
+	}
+	return buffer.Bytes(), hookError
+}
+
+var errtrackerReport = errtracker.Report
+
+func reportHookFailureOnErrtracker(context *Context, err error) {
+	errmsg := err.Error()
+	dupSig := fmt.Sprintf("%s:%s:%s", context.SnapName(), context.HookName(), err)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: maybe this should start with "hook:" ? 
it's the same but given we computed maybe s/err/errmsg/

File: overlord/hookstate/hookmgr.go [hunk 2]
```diff
@@ -296,16 +332,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+	extra := map[string]string{
+		"HookName": context.HookName(),
+	}
+	if context.setup.IgnoreFail {
+		extra["IgnoreFail"] = "1"
+	}
+	oopsid, err := errtrackerReport(context.SnapName(), errmsg, dupSig, extra)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: shouldn't errmsg be dupSig again, or a variant of dupsig with  a newline:

hook:SNAP:HOOK
ERRMSG

?

File: overlord/hookstate/hookmgr.go [hunk 3]
```diff
@@ -296,16 +332,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+	if err != nil {
+		logger.Noticef("cannot report hook failure: %s", err)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: I think we used Debugf for this in snapmgr.go

Also we mostly tend to have uppercase for these "Cannot..."

File: overlord/hookstate/hookmgr.go [hunk 4]
```diff
@@ -296,16 +332,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+	} else {
+		logger.Noticef("reported hook failure from %s as %s", context.HookName(), oopsid)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: "Reported..."

should have the snapName in there?

Also I notice that in the corresponding log message in snapmgr.go for install we don't give a lot of information, maybe we should

File: overlord/hookstate/hookmgr.go [hunk 5]
```diff
@@ -296,16 +333,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+		"HookName": context.HookName(),
+	}
+	if context.setup.IgnoreFail {
+		extra["IgnoreFail"] = "1"
+	}
+	oopsid, err := errtrackerReport(context.SnapName(), errmsg, dupSig, extra)
+	if err != nil {
+		logger.Debugf("Cannot report hook failure: %s", err)
+	} else {
+		logger.Debugf("Reported hook failure from %s for snap %s as %s", context.HookName(), context.SnapName(), oopsid)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: sorry, I think Noticef is fine for this one

also %q is more typical

File: overlord/hookstate/hookmgr.go (hunk 1 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Indeed, I prefixed hook.

File: overlord/hookstate/hookmgr.go (hunk 2 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Errmsg was a bit too terse, I changed that now to something more human readable. If you prefer it to be more symmetrical to dupSig, I can do that as well.

File: overlord/hookstate/hookmgr.go [hunk 6]
```diff
@@ -296,16 +333,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+	if abortOrTimeoutError != nil {
+		return buffer.Bytes(), abortOrTimeoutError
+	}
+	return buffer.Bytes(), hookError
+}
+
+var errtrackerReport = errtracker.Report
+
+func reportHookFailureOnErrtracker(context *Context, output []byte, err error) {
+	errmsg := fmt.Sprintf("hook %s for snap %s failed with: %s. Output: %q", context.HookName(), context.SnapName(), err, output)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: hook %q for snap %q ?

File: overlord/snapstate/snapmgr.go [hunk 7]
```diff
@@ -734,7 +734,7 @@ func (m *SnapManager) undoPrepareSnap(t *state.Task, _ *tomb.Tomb) error {
 	oopsid, err := errtrackerReport(snapsup.SideInfo.RealName, strings.Join(logMsg, "\n"), strings.Join(dupSig, "\n"), extra)
 	st.Lock()
 	if err == nil {
-		logger.Noticef("Reported problem as %s", oopsid)
+		logger.Noticef("Reported problem for %q as %s", snapsup.SideInfo.RealName, oopsid)
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: s/problem/install problem/ ?

File: overlord/hookstate/hookmgr.go (hunk 3 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Thanks, I made them more similar now.

File: overlord/hookstate/hookmgr.go (hunk 4 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Indeed, I added the snapname and fixed the uppercase. Also added the snapname in snamgr.go

File: overlord/hookstate/hookmgr.go [hunk 8]
```diff
@@ -296,16 +333,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+	dupSig := fmt.Sprintf("hook:%s:%s:%s\n%q", context.SnapName(), context.HookName(), err, output)
+	extra := map[string]string{
+		"HookName": context.HookName(),
+	}
+	if context.setup.IgnoreFail {
+		extra["IgnoreFail"] = "1"
+	}
+	oopsid, err := errtrackerReport(context.SnapName(), errmsg, dupSig, extra)
+	if err == nil {
+		logger.Noticef("Reported hook failure from %s for snap %s as %s", context.HookName(), context.SnapName(), oopsid)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: from %q for snap %q?

File: overlord/snapstate/snapmgr.go (hunk 7 above)
Comment (mvo5) [inline overlord/snapstate/snapmgr.go]: Thanks, fixed.

File: overlord/hookstate/hookmgr.go (hunk 8 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Also fixed - good point!

File: overlord/hookstate/hookmgr.go [hunk 9]
```diff
@@ -73,7 +76,10 @@ type HookSetup struct {
 	Hook     string        `json:"hook"`
 	Optional bool          `json:"optional,omitempty"`
 
-	IgnoreFail bool `json:"ignore-fail,omitempty"`
+	IgnoreFail bool          `json:"ignore-fail,omitempty"`
+	MaxRuntime time.Duration `json:"max-runtime,omitempty"`
+
+	ReportOnErrtracker bool `json:"report-on-errtracker,omitempty"`
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: `TrackError bool`

File: overlord/configstate/tasksets.go [hunk 10]
```diff
@@ -37,10 +38,13 @@ func init() {
...
 	hooksup := &hookstate.HookSetup{
-		Snap:       snapName,
-		Hook:       "configure",
-		Optional:   len(patch) == 0,
-		IgnoreFail: flags&snapstate.IgnoreHookFailure != 0,
+		Snap:               snapName,
+		Hook:               "configure",
+		Optional:           len(patch) == 0,
+		IgnoreFail:         flags&snapstate.IgnoreHookFailure != 0,
+		ReportOnErrtracker: flags&snapstate.ReportHookFailureOnErrtracker != 0,
```
Comment (niemeyer) [inline overlord/configstate/tasksets.go]: We can probably simplify the language a bit by dropping analogous terms.. report & track.. failure & error..

`TrackError: ... snapstate.TrackHookError`

Comment (niemeyer) [inline overlord/configstate/tasksets.go]: Ideally the one above would also be:

`IgnoreError:  ... snapstate.IgnoreHookError`,

but unrelated to this PR.

File: overlord/hookstate/hookmgr.go [hunk 11]
```diff
@@ -204,10 +210,14 @@ func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 	if hookExists {
 		output, err := runHook(context, tomb)
 		if err != nil {
+			if hooksup.ReportOnErrtracker {
+				reportHookFailureOnErrtracker(context, output, err)
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: `trackHookError`

File: overlord/hookstate/hookmgr.go [hunk 12]
```diff
@@ -296,16 +333,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+	if abortOrTimeoutError != nil {
+		return buffer.Bytes(), abortOrTimeoutError
+	}
+	return buffer.Bytes(), hookError
+}
+
+var errtrackerReport = errtracker.Report
+
+func reportHookFailureOnErrtracker(context *Context, output []byte, err error) {
+	errmsg := fmt.Sprintf("hook %s for snap %s failed with: %s. Output: %s", context.HookName(), context.SnapName(), err, output)
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: `errmsg := fmt.Sprintf("hook %s in snap %q failed: %v", osutil.OutputErr(output, err))`

File: overlord/hookstate/hookmgr.go [hunk 13]
```diff
@@ -296,16 +333,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
...
+	dupSig := fmt.Sprintf("hook:%s:%s:%s\n%s", context.SnapName(), context.HookName(), err, output)
+	extra := map[string]string{
+		"HookName": context.HookName(),
+	}
+	if context.setup.IgnoreFail {
+		extra["IgnoreFail"] = "1"
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: IgnoreError if we're tweaking it soon, so data already has the new name.

File: overlord/hookstate/hookmgr.go (hunk 9 above)
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: (and "track-error")

Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Thanks, nicer an shorter.

File: overlord/hookstate/hookmgr.go (hunk 11 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: fixed

File: overlord/hookstate/hookmgr.go (hunk 12 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Fixed.

File: overlord/hookstate/hookmgr.go (hunk 13 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Indeed, updated.

File: overlord/configstate/tasksets.go (hunk 10 above)
Comment (mvo5) [inline overlord/configstate/tasksets.go]: Also fixed.

File: overlord/hookstate/hookmgr.go [hunk 14]
```diff
@@ -73,7 +76,9 @@ type HookSetup struct {
 	Hook     string        `json:"hook"`
 	Optional bool          `json:"optional,omitempty"`
 
-	IgnoreFail bool `json:"ignore-fail,omitempty"`
+	Timeout     time.Duration `json:"timeout,omitempty"`
+	IgnoreError bool          `json:"ignore-fail,omitempty"`
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: that should be ignore-error in json too?

Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Fixed

**Pull request merged**